web/auth.go

```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
+	// 	log.Println("LOGIN ERROR: ", err, " USER: ", nuser)
+	// 	data.LoginMessage = "Could not Sign you in"
+	// 	tmp := GetTemplates().Lookup("signin_signup.html")
+	// 	tmp.Execute(w, data)
+	// 	return
+	// }
+
+	// log.Println("USER: ", nuser)
+	// data.LoginMessage = "You are logged in"
+	// tmp := GetTemplates().Lookup("signin_signup.html")
+	// tmp.Execute(w, data)
+}
+
+// SignupHandler handles the signup process for in app signup
+func SignupHandler(w http.ResponseWriter, r *http.Request) {
+	// I feel SignupPageHandler() and this function should share the same struct for better memory conservation
+	data := struct {
+		FBAuthURL     string
+		GoogleAuthURL string
+		LoginMessage  string
+		SignupError   string
+	}{}
+
+	data.LoginMessage = ""
+	data.SignupError = ""
+
+	// This is set here so that when there are any errors from the signup process,
+	// the link will be passed to the template alongside the errors.
+	data.FBAuthURL = FBOauthConf.AuthCodeURL("state", oauth2.AccessTypeOffline)
+
+	r.ParseForm()
+
+	fullName := r.FormValue("name")
+	email := r.FormValue("email")
+	passwd := r.FormValue("passwd")
+
+	// Check if the Post data not empty and validate them.
+	if fullName == "" && email == "" && passwd == "" {
+
+		data.SignupError = "Please Fill out all required Fields"
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
+		return
+
+	}
+
+	// encrypt password.
+	password, _ := bcrypt.GenerateFromPassword([]byte(passwd), bcrypt.DefaultCost)
+	now := time.Now()
+
+	// create the user data
+	user := models.User{
+		Name:                 fullName,
+		Email:                email,
+		DateCreated:          now,
+		FormattedDateCreated: now.String(),
+		Password:             password,
+	}
+
+	// CHECK IF THE EMAIL HAS ALREADY BEEN USED BEFORE.
+	_, err := user.CheckUser(config.GetConf())
+	if err == nil {
+		data.SignupError = "Email address has already been used."
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
+		return
+	}
+	// *** Upsert is replacing documents on the collection. Does not insert new ones***
+
+	// Upsert the user data to the db
+	// err = user.Upsert(config.GetConf())
+	// if err != nil {
+	// 	log.Println(err)
+	// 	data.SignupError = "Could not Sign you up rigt now. Try Again"
+	// 	tmp := GetTemplates().Lookup("signin_signup.html")
+	// 	tmp.Execute(w, data)
+	// 	return
+	// }
+	err = user.Insert(config.GetConf())
+	if err != nil {
+		data.SignupError = "Could not sign you up. Try again."
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
+		return
+	}
+	// note: Userget(r) is passing a string to User.Password instead of []byte
+
+	// SHOULD VERIFY EMAIL ADDRESS SENT, HERE.
+
+	// If all goes well, generate a token for the cookie
+	loginResp, err := GenerateJWT(user)
+	if err != nil {
+		log.Println(err)
+		http.Redirect(w, r, "/signup", 301)
+		return
+	}
+	// Create cookie
+	expire := time.Now().AddDate(0, 0, 1)
```

Comment:
The same token should be used. Its basically an encoded json object containing the users data (the struct)